### PR TITLE
Add new constructor for Domain

### PR DESCRIPTION
**Pull request by mar-celine**
## Proposed changes

Adds a constructor such that the neighbors/orientations can be deduced without the corners being provided.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Domain/Domain.hpp [hunk 1]
```diff
@@ -39,6 +39,25 @@ class Domain {
  public:
   explicit Domain(std::vector<Block<VolumeDim>> blocks) noexcept;
 
+  /*!
+   * Create a Domain using CoordinateMaps to encode the Orientations,
```
Comment (nilsdeppe) [inline src/Domain/Domain.hpp]: should there be a period instead of a comma at the end of this line?

File: src/Domain/Domain.hpp [hunk 2]
```diff
@@ -39,6 +39,25 @@ class Domain {
...
+   *
+   * \details A constructor that does not rely on a corner numbering scheme
+   * being provided by the user in order to set up the Orientations between
+   * blocks. Constructs a global corner numbering for each pair
+   * of abutting Blocks from their maps alone. The numbering is used to
+   * set up the corresponding Orientation, and then is discarded; the
+   * next pair of blocks uses a new global corner numbering, and so on,
+   * until all pairs of abutting Blocks have had their Orientations
+   * determined. For more information on setting up domains, see the
+   * [domain creation tutorial](@ref tutorial_domain_creation).
```
Comment (nilsdeppe) [inline src/Domain/Domain.hpp]: `@ref` -> `\ref` since we've been using the `\` syntax

File: src/Domain/DomainHelpers.cpp [hunk 3]
```diff
@@ -401,6 +461,39 @@ void set_internal_boundaries(
   }
 }
 
+template <size_t VolumeDim>
+void set_internal_boundaries(
+    const std::vector<std::unique_ptr<domain::CoordinateMapBase<
+        Frame::Logical, Frame::Inertial, VolumeDim>>>& maps,
+    gsl::not_null<
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.cpp]: `const` and we prefer `notnull, std::vector` order, though I think that's inconsistent in the existing function of this name so I'm fine filling an issue and addressing it in both functions later

File: src/Domain/DomainHelpers.cpp [hunk 4]
```diff
@@ -401,6 +461,39 @@ void set_internal_boundaries(
...
+          corners_from_two_maps(maps[block1_index], maps[block2_index]);
+      if (block1_index != block2_index and
+          get_common_global_corners<VolumeDim>(corners_for_blocks1_and_2[0],
+                                               corners_for_blocks1_and_2[1])
+                  .size() == two_to_the(VolumeDim - 1)) {
+        const auto orientation_helper =
+            obtain_correspondence_between_blocks<VolumeDim>(
+                0, 1, corners_for_blocks1_and_2);
+
+        // obtain_correspondence_between_maps<VolumeDim>(block1_index,
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.cpp]: can this commented out code be removed?

File: src/Domain/DomainHelpers.cpp [hunk 5]
```diff
@@ -261,6 +261,66 @@ obtain_correspondence_between_blocks(
...
+    }
+  }
+  return {corners_for_block1, corners_for_block2};
+}
+
+template <size_t VolumeDim>
+std::pair<std::array<Direction<VolumeDim>, VolumeDim>,
+          std::array<Direction<VolumeDim>, VolumeDim>>
+obtain_correspondence_between_maps(
+    size_t map_id1, size_t map_id2,
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.cpp]: `const size_t` x 2

File: src/Domain/DomainHelpers.cpp [hunk 6]
```diff
@@ -261,6 +261,66 @@ obtain_correspondence_between_blocks(
...
+              vci_map1.coords_of_corner()});
+      double max_separation = 0.0;
+      for (size_t j = 0; j < VolumeDim; j++) {
+        max_separation =
+            std::max(max_separation,
+                     std::abs(mapped_coords2.get(j) - mapped_coords1.get(j)));
+      }
+      // If true, then the mapped_coords lie on top of one another.
+      // This corner of map2 is assigned the same number as that of map1.
+      if (max_separation < 1.0e-6) {
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.cpp:300]: Why `1.0e-6`? Could this be passed in as an argument to the function? Could `equal_within_roundoff` be used here? My preferred setup, if possible, would be taking `epsilon` and `scale` as arguments, then passing them to `equal_within_roundoff` here. It seems like ideally the arguments would be specifiable in `set_internal_boundaries` so each domain creator can adjust them as necessary.

File: src/Domain/DomainHelpers.cpp [hunk 7]
```diff
@@ -261,6 +261,66 @@ obtain_correspondence_between_blocks(
...
+        gsl::at(corners_for_block2, vci_map2.local_corner_number()) =
+            vci_map1.local_corner_number();
+        break;
+      }
+      // Otherwise, a new number is assigned to this corner.
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.cpp]: We generally prefer this comment be inside the `else {` block

File: src/Domain/DomainHelpers.cpp [hunk 8]
```diff
@@ -261,6 +261,66 @@ obtain_correspondence_between_blocks(
...
+  for (VolumeCornerIterator<VolumeDim> vci{}; vci; ++vci) {
+    gsl::at(corners_for_block1, vci.local_corner_number()) =
+        vci.local_corner_number();
+  }
+  // Fill corners_for_block2 based off which corners are shared with map1.
+  for (VolumeCornerIterator<VolumeDim> vci_map2{}; vci_map2; ++vci_map2) {
+    const auto mapped_coords2 =
+        (*map2)(tnsr::I<double, VolumeDim, Frame::Logical>{
+            vci_map2.coords_of_corner()});
+    double num_unshared_corners = 0;
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.cpp]: why a `double` and not a `size_t`?

File: src/Domain/DomainHelpers.cpp [hunk 9]
```diff
@@ -261,6 +261,66 @@ obtain_correspondence_between_blocks(
...
+      if (max_separation < 1.0e-6) {
+        gsl::at(corners_for_block2, vci_map2.local_corner_number()) =
+            vci_map1.local_corner_number();
+        break;
+      }
+      // Otherwise, a new number is assigned to this corner.
+      else {
+        gsl::at(corners_for_block2, vci_map2.local_corner_number()) =
+            two_to_the(VolumeDim) + num_unshared_corners;
+        num_unshared_corners++;
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.cpp:315]: What's not clear to me is how if there are many blocks you don't get duplicate corner numbers since `num_unshared_corners` always starts at zero. Does that not matter because you just treat each pair of blocks independent of all the others? I.e. you don't create a global corner numbering? If so, I think having a comment along the lines of:
```
// Set num_unshared_corners to zero because we only need a local corner numbering scheme. 
// This is because our algorithm treats each pair of blocks independently to create the 
// orientation (if any) between them.
```
where `num_unshared_corners` is created would be useful

File: src/Domain/DomainHelpers.hpp [hunk 10]
```diff
@@ -61,6 +61,19 @@ void set_internal_boundaries(
         std::vector<DirectionMap<VolumeDim, BlockNeighbor<VolumeDim>>>*>
         neighbors_of_all_blocks) noexcept;
 
+/// \ingroup ComputationalDomainGroup
+/// Sets up the BlockNeighbors using the corner numbering scheme
+/// implied by the maps provided by the user to deduce the correct
+/// neighbors and orientations. Does not set up periodic boundary
```
Comment (nilsdeppe) [inline src/Domain/DomainHelpers.hpp]: This is a long "brief" comment. Perhaps split the second sentence off and making it a `\warning` since that's a fairly important thing to be aware of :)

File: tests/Unit/Domain/Creators/Test_Sphere.cpp [hunk 11]
```diff
@@ -33,8 +33,11 @@
 #include "Framework/TestHelpers.hpp"
 #include "Helpers/Domain/DomainTestHelpers.hpp"
 #include "Parallel/RegisterDerivedClassesWithCharm.hpp"
+#include "Utilities/CloneUniquePtrs.hpp"
 #include "Utilities/MakeArray.hpp"
 
+#include <iostream>
```
Comment (nilsdeppe) [inline tests/Unit/Domain/Creators/Test_Sphere.cpp]: Oops, I think the `iostream` include is left over from debugging

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
This is really nice @gsb76! Thank you for implementing this :) A few small changes, please feel free to squash immediately :)

File: src/Domain/DomainHelpers.cpp (hunk 6 above)
Comment (mar-celine) [reply src/Domain/DomainHelpers.cpp:300]: I'm not sure about changing this, because so far we've used analytic maps that are constructed to agree at the boundaries. If `max_separation` is above any small number, the maps are most likely disconnected, and if they are connected, they share an analytic map at the boundary so the `max_separation` is likely very small. I chose `1.0e-6` because so far all separations in the domains I've tested have been clearly on one side of this bound or the other. To have an argument be passed in suggests the map connectivity will be "fuzzy" where they're only connected up to some tolerance.

An alternative I would suggest, if desired, is an additional function that finds the `max_separation` in a domain where the maps are properly connected, that would be a good diagnostic in some cases.

File: src/Domain/DomainHelpers.cpp (hunk 9 above)
Comment (mar-celine) [reply src/Domain/DomainHelpers.cpp:315]: That is correct :smile: there is never a global numbering created, since relative orientations only care about two blocks at a time.

Comment (nilsdeppe) [reply src/Domain/DomainHelpers.cpp:315]: Sweet :D Thank you!

File: src/Domain/DomainHelpers.cpp (hunk 6 above)
Comment (nilsdeppe) [reply src/Domain/DomainHelpers.cpp:300]: I guess my question is more if I rescale the entire domain radially by 1.0e8 then `1.0e-6` may no longer be large enough to check agreement. It seems like the separation is, even for analytic maps, a relative quantity compared to some measure of the location of the boundary/size of the domain.  E.g. if we took the BBH domain and just made the holes be 1e10 solar masses then the separation between domains should probably be scaled according to that size. I'm not sure if we'll encounter this in practice. I'm just asking to see what you're thoughts are :)

File: src/Domain/DomainHelpers.hpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Domain/DomainHelpers.hpp]: Did you want to do this? Don't feel like you have to, just a suggestion :)

**Comment by nilsdeppe** [comment]
LGTM, I'm happy for you to squash everything as you see fit, including if you decide you want to address the ping :)

File: src/Domain/DomainHelpers.cpp (hunk 6 above)
Comment (mar-celine) [reply src/Domain/DomainHelpers.cpp:300]: I agree - ideally the threshold is some size relative to the size of the maps and I've assumed that in practice all maps have a size of order ~1 at least to within a couple of orders of magnitude. I can add a comment and/or error message that discusses this, because I think if we do encounter this it will be in special cases only. :+1:

File: src/Domain/DomainHelpers.hpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Domain/DomainHelpers.hpp]: Sorry, I missed this...

File: src/Domain/DomainHelpers.cpp (hunk 6 above)
Comment (nilsdeppe) [reply src/Domain/DomainHelpers.cpp:300]: Okay, sounds good!

**Comment by nilsdeppe** [comment]
Okay, LGTM please go ahead and squash :)

File: src/Domain/Domain.hpp [hunk 12]
```diff
@@ -39,6 +39,25 @@ class Domain {
  public:
   explicit Domain(std::vector<Block<VolumeDim>> blocks) noexcept;
 
+  /*!
+   * Create a Domain using CoordinateMaps to encode the Orientations.
+   * This constructor does not support periodic boundary conditions.
+   *
+   * \details A constructor that does not rely on a corner numbering scheme
+   * being provided by the user in order to set up the Orientations between
+   * blocks. Constructs a global corner numbering for each pair
```
Comment (markscheel) [inline src/Domain/Domain.hpp]: Maybe reword "A constructor that does not require the user to provide a corner numbering scheme. Constructs a global corner numbering..."

**Review by markscheel — COMMENTED** [review]
Just one possible clarification of a comment, otherwise looks good!

**Review by markscheel — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]
